### ISSUE TITLE
DockerClientFactory#client() dead lock when running docker inside docker (DiD)

### ISSUE THREAD

**Issue by radistao** [testcontainers/testcontainers-java#1602]
`DockerClientFactory#client()` causes deadlock when running docker-in-docker and tries to resolve IP address.

### Environment
 _docker-in-docker_ (gradle build in Jenkins)

### When
resolving `DockerClientProviderStrategy#getDockerHostIpAddress()`

**Description and steps to reproduce**:
We use [`PostgreSQLContainer` 1.11.4](https://mvnrepository.com/artifact/org.testcontainers/postgresql/1.11.4) to run DB instance: 
  1. `new PostgreSQLContainer();`

  1. It calls parent constructors till `GenericContainer()` initializing [`DockerClientFactory#client()`](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/containers/GenericContainer.java#L138) singleton instance.

   1. `DockerClientFactory#client()` uses [`synchronized(this.$lock)`](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/DockerClientFactory.java#L95):
![image](https://user-images.githubusercontent.com/580901/60980636-23291580-a335-11e9-9854-81f294dd5bb5.png)

  1. Tries to [get host ip](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/DockerClientFactory.java#L107)

  1. [the host URI is `unix:///var/run/docker.sock`, scheme is `unix` and running `IN_A_CONTAINER`](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/dockerclient/DockerClientConfigUtils.java#L57) => goto [`DockerClientConfigUtils#getDefaultGateway()`](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/dockerclient/DockerClientConfigUtils.java#L25)

  1. To define the gateway they want run [`ip route|awk '/default/ { print $3 }'` inside `alpine` container](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/dockerclient/DockerClientConfigUtils.java#L27). To achieve this they need to call [DockerClientFactory#runInsideDocker(Consumer, BiFunction)](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/dockerclient/DockerClientConfigUtils.java#L26)

  1. **[`DockerClientFactory#runInsideDocker(Consumer, BiFunction)`](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/DockerClientFactory.java#L234) fetches (initializes) [_client_ instance again, because _strategy_ is still null!](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/DockerClientFactory.java#L234)**
![image](https://user-images.githubusercontent.com/580901/60982155-e14d9e80-a337-11e9-937e-3edc7f6789b1.png)

  1. **call `DockerClientFactory#client()`, but it is synchronized and locked on STEP 3 -> DEAD LOCK!!**

### Environment:
Testcontainers Version: [**1.11.4**](https://mvnrepository.com/artifact/org.testcontainers/testcontainers/1.11.4)
JVM: `11.0.3 (Azul Systems, Inc. 11.0.3+7-LTS)`
Gradle: `5.5`
Platform:  
```
Client:
  Version:           18.09.6
  API version:       1.39
  Built:             Sat May  4 02:35:27 2019
  OS/Arch:           linux/amd64

Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.12)
  Built:            Sun Feb 10 03:42:13 2019
  OS/Arch:          linux/amd64
````
### Notes:

This works correct when running in host systems, because in this case `IN_A_CONTAINER` is false and `getDockerHostIpAddress()` is resolved to default _"localhost"_ value.

**Comment by bsideup** [comment]
Hi @radistao,

The strategy field is initialized to `getFirstValidStrategy` before calling `getDockerHostIpAddress`, and `runInsideDocker` will not call `this.client()`, but use `strategy.getClient()`.

Are you sure that's what you observe?

**Comment by bsideup** [comment]
P.S. we do run Docker-in-Docker tests during our CI:
https://github.com/testcontainers/testcontainers-java/blob/d681aeef6f0d2987b6126dbc7382d6eb83fb7f3b/.travis.yml#L38

**Comment by radistao** [comment]
you are right, i made wrong assumption about  _strategy_. I'll check deeper.

I definitely have the issue running in our Jenkins: the tests timeout when i call either `new PostgreSQLContainer()` or even just `DockerClientFactory.instance().client()`.

It's hard to test or debug, because reproduced only on Jenkins build.
The last logged lines are:

> INFO  o.t.d.DockerClientProviderStrategy - Loaded org.testcontainers.dockerclient.EnvironmentAndSystemPropertyClientProviderStrategy from ~/.testcontainers.properties, will try it first
> INFO  o.t.d.DockerClientProviderStrategy - Will use 'okhttp' transport
> DEBUG o.t.d.DockerClientProviderStrategy - Pinging docker daemon...
> DEBUG c.g.d.c.c.AbstrDockerCmd - Cmd: org.testcontainers.dockerclient.transport.okhttp.OkHttpDockerCmdExecFactory$1@64ae2b26
> INFO  o.t.d.EnvironmentAndSystemPropertyClientProviderStrategy - Found docker client settings from environment
> INFO  o.t.d.DockerClientProviderStrategy - Found Docker environment with Environment variables, system properties and defaults. Resolved: 
>     dockerHost=unix:///var/run/docker.sock
>     apiVersion='{UNKNOWN_VERSION}'
>     registryUrl='https://index.docker.io/v1/'
>     registryUsername='root'
>     registryPassword='null'
>     registryEmail='null'
>     dockerConfig='DefaultDockerClientConfig[dockerHost=unix:///var/run/docker.sock,registryUsername=root,registryPassword=<null>,registryEmail=<null>,registryUrl=https://index.docker.io/v1/,dockerConfigPath=/company/jenkins/workspace/XXXX/abc.defgg.service.our-service/custom-build-image-7c7384d3-cbd4-4a52-818d-4f9cd3398d9d@tmp/756e1a53-6fe5-4fdf-bd86-aa285ecfb5dc,sslConfig=<null>,apiVersion={UNKNOWN_VERSION},dockerConfig=<null>]'
> 
> DEBUG c.g.d.c.c.AbstrDockerCmd - Cmd: ListImagesCmdImpl[imageNameFilter=alpine:3.5,showAll=false,filters=com.github.dockerjava.core.util.FiltersBuilder@0,execution=com.github.dockerjava.core.exec.ListImagesCmdExec@4ed5ee81]
> DEBUG o.t.u.RegistryAuthLocator - Looking up auth config for image: alpine:3.5
> DEBUG o.t.u.RegistryAuthLocator - RegistryAuthLocator has configFile: /company/jenkins/workspace/XXXX/abc.defgg.service.our-service/custom-build-image-7c7384d3-cbd4-4a52-818d-4f9cd3398d9d@tmp/756e1a53-6fe5-4fdf-bd86-aa285ecfb5dc/config.json (exists) and commandPathPrefix: 
> DEBUG o.t.u.RegistryAuthLocator - registryName [index.docker.io] for dockerImageName [alpine:3.5]
> DEBUG o.t.u.RegistryAuthLocator - no matching Auth Configs - falling back to defaultAuthConfig [null]
> DEBUG o.t.d.a.AuthDelegatingDockerClientConfig - Effective auth config [null]
> DEBUG c.g.d.c.c.AbstrDockerCmd - Cmd: com.github.dockerjava.core.command.CreateContainerCmdImpl@7cd2a591[name=<null>,hostName=<null>,domainName=<null>,user=<null>,attachStdin=<null>,attachStdout=<null>,attachStderr=<null>,portSpecs=<null>,tty=<null>,stdinOpen=<null>,stdInOnce=<null>,env=<null>,cmd={sh,-c,ip route|awk '/default/ { print $3 }'},entrypoint=<null>,image=alpine:3.5,volumes=com.github.dockerjava.api.model.Volumes@bbbfa95,workingDir=<null>,macAddress=<null>,networkDisabled=<null>,exposedPorts=com.github.dockerjava.api.model.ExposedPorts@622c285e,stopSignal=<null>,hostConfig=com.github.dockerjava.api.model.HostConfig@6744b7a0[binds=<null>,blkioWeight=<null>,blkioWeightDevice=<null>,blkioDeviceReadBps=<null>,blkioDeviceReadIOps=<null>,blkioDeviceWriteBps=<null>,blkioDeviceWriteIOps=<null>,memorySwappiness=<null>,capAdd=<null>,capDrop=<null>,containerIDFile=<null>,cpuPeriod=<null>,cpuShares=<null>,cpuQuota=<null>,cpusetCpus=<null>,cpusetMems=<null>,devices=<null>,diskQuota=<null>,dns=<null>,dnsSearch=<null>,extraHosts=<null>,links=<null>,logConfig=<null>,lxcConf=<null>,memory=<null>,memorySwap=<null>,memoryReservation=<null>,kernelMemory=<null>,networkMode=<null>,oomKillDisable=<null>,autoRemove=<null>,oomScoreAdj=<null>,portBindings=<null>,privileged=<null>,publishAllPorts=<null>,readonlyRootfs=<null>,restartPolicy=<null>,ulimits=<null>,volumesFrom=<null>,pidMode=<null>,securityOpts=<null>,cgroupParent=<null>,volumeDriver=<null>,shmSize=<null>,pidsLimit=<null>,runtime=<null>,tmpFs=<null>],labels={org.testcontainers=true, org.testcontainers.sessionId=edc44e8c-c108-4684-8761-77dbe2d4d232},networkingConfig=<null>,ipv4Address=<null>,ipv6Address=<null>,aliases=<null>,authConfig=<null>,execution=com.github.dockerjava.core.exec.CreateContainerCmdExec@40e51001]
> DEBUG c.g.d.c.c.AbstrDockerCmd - Cmd: 55e9d159b1cac3f92d95ec912d5cea5bb4dd35750151f0a2ed4b68481eb1713b,com.github.dockerjava.core.exec.StartContainerCmdExec@20cb938b
> DEBUG c.g.d.c.c.LogContainerResultCallback - STDOUT: 172.17.0.1
> 

Then after couple of minutes the build is interrupted by timeout.

To run the build container i use:
`args '-v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock -v $PWD:$PWD -w $PWD -u root --privileged'`

I'll investigate tomorrow deeper and report the result.

**Comment by radistao** [comment]
closing it so far.
if i find more details - i'll reopen it.

**Comment by radistao** [comment]
@bsideup 
i was able to trace till first output from the container:

>11:35:34.426 [Test worker] TRACE c.g.d.c.e.LogContainerCmdExec - GET: OkHttpWebTarget(okHttpClient=org.testcontainers.shaded.okhttp3.OkHttpClient@7974109c, baseUrl=http://docker.socket/, path=[/containers/2a11e9f094a7b65b712d7527abc01099f12e3b7ad60860d8aecd8d0fe0e64fff/logs], queryParams={stdout=[true], follow=[true]})
> 
>11:35:34.438 [tc-okhttp-stream-1352362980] DEBUG c.g.d.c.c.LogContainerResultCallback - STDOUT: 172.17.0.1

so, it fetches the ip address from the container output (`ip route|awk '/default/ { print $3 }'`), but than "hangs", not sure where.
That last successful output is from `LogContainerResultCallback#onNext()`. 
It seems it is stuck in the infinite or blocked loop of `FrameSink` trying to fetch more from the output:
https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/dockerclient/transport/okhttp/OkHttpInvocationBuilder.java#L360

Any ideas what could cause this blocking or how to trace/debug deeper?
Thank you.

**Comment by jannis-baratheon** [comment]
Usually situations like this when the app suddenly stops/hangs are related to threads dying silently without a callback to the parent thread.

Wild guess: maybe [this thread here](https://github.com/testcontainers/testcontainers-java/blob/1.11.4/core/src/main/java/org/testcontainers/dockerclient/transport/okhttp/OkHttpInvocationBuilder.java#L304) dies with an unhandled exception and therefore fails to call any of the termination methods (`onComplete`, `onError`).

You could try and set a global unhandled exception handler with [Thread.html#setDefaultUncaughtExceptionHandler](https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.UncaughtExceptionHandler.html) and add some logging.

**Comment by radistao** [comment]
Thank you, jannis-baratheon.
It is very likely smth wrong inside that thread, because it is traced only once.
Unfortunately, nothing caught in the `defaultUncaughtExceptionHandler` which i set...

**Comment by jannis-baratheon** [comment]
What about debugging @radistao?

**Comment by radistao** [comment]
i don't know how to connect to docker in docker in remote Jenkins server, which is likely doesn't open any debugging ports :)

**Comment by jannis-baratheon** [comment]
It's likely that an [IOError](https://docs.oracle.com/javase/8/docs/api/java/io/IOError.html) gets thrown somewhere and from what I can see only `Exception`s are handled in `OkHttpInvocationBuilder`.

I'd recommend finding a way to debug into the process. If you have access to the server just try running your tests there with debugger attached. Another way would be to prepare a custom Testcontainers build with logging added in crucial places, then push it to the CI and investigate the output then.

**Comment by radistao** [comment]
i overridden `OkHttpInvocationBuilder` so that it catches all possible Throwable and traces every line in `OkHttpInvocationBuilder.FramedSink#accept()`:
```java
while (true) {
    try {
        logger.info("source={}, source.exhausted={}", source, source.exhausted());
        if (!source.exhausted()) {
            logger.info("source request");
            if (!source.request(8L)) {
                logger.info("source request=false");
                return;
            }
            logger.info("source request=true");

            logger.info("source request: 1");
            StreamType streamType = streamType(source.readByte());
            logger.info("source request: 2");
            source.skip(3L);
            logger.info("source request: 3");
            int payloadSize = source.readInt();
            logger.info("source request: 4");
            if (streamType != StreamType.RAW) {
                logger.info("source request: 5");
                if (!source.request((long) payloadSize)) {
                    logger.info("source request: 6");
                    return;
                }

                logger.info("source request: 7");
                byte[] payload = source.readByteArray((long) payloadSize);
                logger.info("source request: 8");
                this.resultCallback.onNext(new Frame(streamType, payload));
                logger.info("source request: 9");
                continue;
            }

            logger.info("this.resultCallback.onNext");
            this.resultCallback.onNext(new Frame(streamType, source.readByteArray()));
            logger.info("this.resultCallback.onNext after");
            continue;
        }
    } catch (Throwable var5) {
        this.resultCallback.onError(var5);
    }

    logger.info("accept finished");
    return;
}
```
The logs i see now are very strange!
The last logged line is _source request: 9_, which means it successfully captured the output (and traced it as _STDOUT: 172.17.0.1_), and now it should `continue` to the beginning of the `while(true)` loop.

BUT the first line in the loop (_source={}, source.exhausted={}_) is not logged any more and the application/tread hangs at this point _in the edge of the loop_ ...

**Comment by bsideup** [comment]
@radistao this can only mean that `source.exhausted()` blocks forever o_O

**Comment by radistao** [comment]
yes, i already found this. Trying now to log `org.testcontainers.shaded.okhttp3.internal.connection.Exchange.ResponseBodySource#read()`

**Comment by jannis-baratheon** [comment]
> [jvm](https://square.github.io/okio/2.x/okio/jvm/index.html) / [okio](https://square.github.io/okio/2.x/okio/jvm/okio/index.html) / [BufferedSource](https://square.github.io/okio/2.x/okio/jvm/okio/-buffered-source/index.html) / [exhausted](https://square.github.io/okio/2.x/okio/jvm/okio/-buffered-source/exhausted.html)  
> 
> # exhausted
> 
> <a name="okio.BufferedSource$exhausted()" href=""></a>`abstract fun exhausted(): [Boolean](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-boolean/index.html)`
> 
> Returns true if there are no more bytes in this source. This will block until there are bytes to read or the source is definitely exhausted.
>
> -- https://square.github.io/okio/2.x/okio/jvm/okio/-buffered-source/exhausted.html

**Comment by jannis-baratheon** [comment]
Probably stating the obvious here but looks like the command doesn't terminate.

**Comment by jannis-baratheon** [comment]
Shooting in the dark here, but please try fetching standard error:

```
final LogContainerCmdImpl logContainerCmd2 = (LogContainerCmdImpl) strategyDockerClient
        .logContainerCmd(createContainerCmdId)
        .withStdOut(true)
        **.withStdErr(true)**
        .withFollowStream(true);
```

**Comment by radistao** [comment]
unfortunately i'll be able to continue only on Monday.

**Comment by jannis-baratheon** [comment]
@radistao Apart from what I wrote above I'd also recommend preparing a test not using Testcontainers/DockerClient. Just add a test which invokes the docker command directly using `ProcessBuilder` or `Runtime.getRuntime.exec()` and the reads the output (both stderr and stdout) like here: https://stackoverflow.com/a/23322785/4494577

We'll see if it's something in the libs or docker itself.

**Comment by radistao** [comment]
`withStdErr()` didn't show any info.

will try using `Runtime#exec()`.

BTW, if i run `docker run --rm alpine:latest sh -c ip route|awk '/default/ { print  $3 }'` in that Jenkins shell (without before gradle build) - it works properly and prints IP address.

**Comment by radistao** [comment]
@jannis-baratheon 
tried this:
```java
String standardOutput;
try {
    final Process proc = new ProcessBuilder("docker", "run", "--rm", "alpine:latest", "sh",
            "-c", "ip route|awk '/default/ { print $3 }'")
            .redirectOutput(ProcessBuilder.Redirect.PIPE)
            .redirectError(ProcessBuilder.Redirect.PIPE)
            .redirectErrorStream(true)
            .start();

    proc.waitFor(1, MINUTES);
    standardOutput = new String(proc.getInputStream().readAllBytes());
} catch (Throwable e) {
    standardOutput = "Exception: " + e.getMessage();
}

logger.info("DOCKER RUN IN SHELL: [{}]", standardOutput);
```

works fine and prints:
>[172.17.0.1
>]

(with trailing new line, if it is important)

**Comment by radistao** [comment]
also i was able to trace/debug till this blocking line:

`UnixDomainSocket#doRead(ByteBuffer)`:

```java
UnixDomainSocketLibrary.read(var2, var1, var1.remaining());
```

Than it goes to _native_ platform calls, which i can't debug any deeper ((

**Comment by jannis-baratheon** [comment]
> (with trailing new line, if it is important)

And locally (where it works) does it add this trailing new line?

**Comment by bsideup** [comment]
@radistao could you please try changing the read timeout to something like 60 seconds?
https://github.com/testcontainers/testcontainers-java/blob/b840c5d77239747995938b9a123e8a5722d89d9e/core/src/main/java/org/testcontainers/dockerclient/transport/okhttp/OkHttpDockerCmdExecFactory.java#L43

**Comment by radistao** [comment]
tried `echo` instead of `route|awk`:
```java
final CreateContainerCmd createContainerCmd = strategyDockerClient
        .createContainerCmd(TINY_IMAGE)
        .withLabels(DEFAULT_LABELS)
        .withCmd("sh", "-c", "echo 172.17.0.1");
```

this also blocks after full output reading (successfully fetches and logs the address, but then blocks)

**Comment by radistao** [comment]
@jannis-baratheon 
>And locally (where it works) does it add this trailing new line?

yes, it prints a new line (you can try at your command line
 `docker run --rm alpine:latest sh -c "ip route|awk '/default/ { print \$3 }'"`)

**Comment by jannis-baratheon** [comment]
> BTW, if i run `docker run --rm alpine:latest sh -c ip route|awk '/default/ { print $3 }'` in that Jenkins shell (before gradle build) - it works properly and prints IP address.
>
> -- https://github.com/testcontainers/testcontainers-java/issues/1602#issuecomment-511379968

But do you call this command from within a docker container on Jenkins like tests do? Something like this:

```
docker run -t --rm \ 
	-v /var/run/docker.sock:/var/run/docker.sock \
	openjdk:8-jdk-alpine \
	-c "docker run --rm alpine:latest sh -c ip route|awk '/default/ { print $3 }'"
```

**Comment by jannis-baratheon** [comment]
> this also blocks after full output reading (successfully fetches and logs the address, but then blocks)
>
> -- https://github.com/testcontainers/testcontainers-java/issues/1602#issuecomment-511385966

And is the `alpine:latest` container running when this happens (`docker ps`)? Or has it exited already. I guess another thing to dig into.

Anyway I guess this confirms that it's neither a Testcontainers nor a DockerClient issue.

**Comment by radistao** [comment]
@bsideup 
> could you please try changing the read timeout to something like 60 seconds?
tried 
```java
OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder()
    .readTimeout(60, TimeUnit.SECONDS)
    .retryOnConnectionFailure(true);
```

same result ((

**Comment by radistao** [comment]
@jannis-baratheon 

I run Jenkinsfile with the following arguments:

```groovy
pipeline {
  agent {
    docker {
      image 'debian:9'
      args '-v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock -v $PWD:$PWD -w $PWD -u root --privileged'
    }
  }

  stages {
    stage('Test') {
      steps {
        sh 'docker ps -a'
        sh "ip route|awk '/default/ { print \$3 }'"

        sh 'docker run --rm alpine:latest sh -c "ip route|awk \'/default/ { print \\$3 }\'"'

        sh 'docker run -t --rm -v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock ' +
                'debian:9 ' +
                'sh -c "docker run -t --rm alpine:latest sh -c \\"ip route | awk \'/default/ { print \\\\\\$3 }\'\\""'
```

(the last one is _docker(alpine:lates)-in-docker(debian:9)-in-docker(Jenkins build)_)

all 3 of them print `172.17.0.1`

Also unfortunately i can't connect to the Jenkins server to verify `docker ps` or `docker logs` during the build process, but when i restart a new build i see a number of containers with _alpine:3.5 "sh -c 'ip route|awk…"_ exited couple of minutes ago (e.g. from previous builds)

**Comment by jannis-baratheon** [comment]
if you are absolutely sure that this pipeline has the same setup that your tests are running in and you don't have access to the Jenkins agent then I'm outta ideas for now.

**Comment by radistao** [comment]
yeah, seems we came to the dead end. Anyway - thank you for support!

**Comment by radistao** [comment]
@jannis-baratheon 
not sure if it still helps:
i verified `docker ps -a` right after the build finished by timeout (it post-build step):

```
IMAGE         COMMAND                  CREATED             STATUS                   
alpine:3.5    "sh -c 'ip route|awk…"   3 minutes ago       Exited (0) 3 minutes ago
```

so, this means the docker containers started and finished correctly, and `UnixDomainSocketLibrary.read()` waits for exited container

**Comment by jannis-baratheon** [comment]
Maybe this is a Docker issue. Compare Docker versions - your local and Jenkins. And when BufferedSource blocks - has it already read the whole output, including the newline? Or does it hang before the newline? 

@radistao

**Comment by jannis-baratheon** [comment]
I've got a feeling we're missing something obvious 🤔

**Comment by radistao** [comment]
> Compare Docker versions - your local and Jenkins. 

locally i have both client and server version `18.09.7`, on Jenkins client is also `18.09.7`, server is `18.09.2`. Don't think this is an issue.

>And when BufferedSource blocks - has it already read the whole output, including the newline?

the code is trimming `(new String(this.payload)).trim()`, i'll try to verify later (now there are some issues on the CI server)

> I've got a feeling we're missing something obvious

Likely yes. BTW i found that if i run locally with `-t` - sometimes it lags. I'm not sure how this _Allocate a pseudo-TTY_ property works, but could it be that the `DockerClient` tries to connect to some tty, but it is (obviously) not available in Jenkins?

**Comment by jannis-baratheon** [comment]
> Likely yes. BTW i found that if i run locally with `-t` - sometimes it lags. I'm not sure how this _Allocate a pseudo-TTY_ property works, but could it be that the `DockerClient` tries to connect to some tty, but it is (obviously) not available in Jenkins?
>
> -- https://github.com/testcontainers/testcontainers-java/issues/1602#issuecomment-512303289

Some nice reading on this:
https://stackoverflow.com/questions/30137135/confused-about-docker-t-option-to-allocate-a-pseudo-tty

**Comment by jannis-baratheon** [comment]
From what I read it might be worth playing with the `-t` parameter.

**Comment by radistao** [comment]
but that was my question: how to configure it in `DockerClient`? I didn't find any reference to _tty_ it in the documentation and source codes.

**Comment by radistao** [comment]
all 4 finished successfully:

```
docker run --rm -v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock docker-registry.goorange.sixt.com/customerorg/jdk-gradle-debian:zulu11.0.3-gradle5.5-git sh -c docker run --rm alpine:latest sh -c "ip route | awk '/default/ { print \$3 }'"
172.17.0.1

docker run --rm -v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock docker-registry.goorange.sixt.com/customerorg/jdk-gradle-debian:zulu11.0.3-gradle5.5-git sh -c docker run -t --rm alpine:latest sh -c "ip route | awk '/default/ { print \$3 }'"
172.17.0.1

docker run -t --rm -v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock docker-registry.goorange.sixt.com/customerorg/jdk-gradle-debian:zulu11.0.3-gradle5.5-git sh -c docker run --rm alpine:latest sh -c "ip route | awk '/default/ { print \$3 }'"
172.17.0.1

docker run -t --rm -v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock docker-registry.goorange.sixt.com/customerorg/jdk-gradle-debian:zulu11.0.3-gradle5.5-git sh -c docker run -t --rm alpine:latest sh -c "ip route | awk '/default/ { print \$3 }'"
172.17.0.1
```

**Comment by jannis-baratheon** [comment]
And what command does your Jenkins use to tun the tests? (sorry if it's somewhere in the previous comments) Does it have a `-t`?

**Comment by radistao** [comment]
`./gradlew test --info`

**Comment by radistao** [comment]
> And when BufferedSource blocks - has it already read the whole output, including the newline?

I debugged that one: it reads completely, including new line (`0x0A` code)

**Comment by jannis-baratheon** [comment]
Sorry I'm not a Jenkins specialist, but are you using some plugin to run docker-in-docker tests?

**Comment by jannis-baratheon** [comment]
Have you seen #1473?

**Comment by jannis-baratheon** [comment]
I guess not only you are experiencing the problem...

**Comment by jannis-baratheon** [comment]
Duplicate of #1473

**Comment by radistao** [comment]
yes, seems similar (sorry, i tried to search before creating an issue, but didn't find).
Does it have sense to copy some comments to that thread? At least i was able to debug till blocking native call `UnixDomainSocketLibrary.read()`.

**Comment by jannis-baratheon** [comment]
No worries, also haven't found it before. I don't think it's a good idea to copy comments there. You can use this ticket as reference when discussing there. I already asked about the progress. You did a lot of investigation work here, hope it will push #1473 forward.

**Comment by jannis-baratheon** [comment]
Anyway my question still holds @radistao. What is your Jenkins setup? How do you spawn containers? Do you use a plugin or maybe some vanilla Jenkins feature?

**Comment by radistao** [comment]
I can't access this build service directly, the only thing i can say it is a [Jenkins Blue Ocean](https://jenkins.io/projects/blueocean/) instance with following pipeline configuration:

```groovy
pipeline {
  agent {
    docker {
      image 'debian:stretch'
      args '-v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock -v $PWD:$PWD -w $PWD -u root --privileged'
    }
  }

  options {
    disableConcurrentBuilds()
    timeout(time: 5, unit: 'MINUTES')
  }

  stages {
    stage('Test') {
      steps {
        sh 'docker version'
        sh 'docker ps -a'
        sh "ip route|awk '/default/ { print \$3 }'"
        sh 'docker run --rm alpine:3.5 sh -c "ip route|awk \'/default/ { print \\$3 }\'"'

        sh 'docker run -t --rm -v /usr/bin/docker:/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock ' +
                'docker-registry.goorange.sixt.com/customerorg/jdk-gradle-debian:zulu11.0.3-gradle5.5-git ' +
                'sh -c "docker run -t --rm alpine:3.5 sh -c \\"ip route | awk \'/default/ { print \\\\\\$3 }\'\\""'

        sh './gradlew testIntegrationClasses' // low logging level

        sh './gradlew testIntegration -x test --info --tests com.sixt.service.identity_management.repository.JdbcUserRepositoryIT.doNothingForTestDockerOnly'

      }
    }
  }
}
```

**Comment by radistao** [comment]
Follow https://github.com/testcontainers/testcontainers-java/issues/1473 for further related discussion.
